design/reboot-interface.md

```diff
@@ -0,0 +1,193 @@
+* The ability to determine a point where all processes running on the Host at
+  the time of the fencing decision are guaranteed to have been stopped
+* Rapid convergence to a state where all running processes are stopped,
+  independent of other happenings in the system
+
+### Non-Goals
+
+There is no requirement to implement a scheduled reboot. In the Kubernetes
+context, reboot decisions should generally be made in realtime by some higher
+level of the Node/Machine hierarchy, to take into account such questions as the
+overall health of the cluster and the effect of a reboot on that. The best
+implementation for this would be a RebootSchedule CRD that waits until the
+appointed time before issuing an immediate reboot request to the BareMetalHost.
+This allows multiple reboots to be scheduled, scheduled reboots to be
+manipulated or cancelled, and a record to be left behind of past scheduled
+reboots. The proposed design could easily be extended to accomodate this
+requirement should it arise in future.
+
+There is no requirement to hold the power off for any particular length of
+time. For fencing purposes we only need to be sure when the node is powered
+off.
+
+This API is not responsible for managing unprovisioned hosts, e.g. to recover
+from state inconsistencies in Ironic prior to provisioning. Any such
+inconsistencies that are not handled internally by the BareMetalHost will
+likely need manual intervention to recover anyway.
+
+## Proposal
+
+A new date-time field, ``lastPoweredOn``, will be added to the ``provisioning``
+section of the BareMetalHost status. This records a time after which the Host
+was last booted using the current image. Processes running prior to this time
+may be assumed to have been stopped.
+
+To provide a declarative interface for performing a reboot, a new
+BareMetalHostRebootRequest CRD will be created. The spec of this custom
+resource will contain only a reference to a BareMetalHost. The status will
+contain timestamps for when the Host was powered off and on again. Once the
+power off time is present in the status, the host has been stopped. Each
+RebootRequest will be owned by the Host it refers to, so that if the Host is
+deleted all associated RebootRequest CRs will be cleaned up.
+
+The actual power management will be performed by the Host controller. This is
+necessary to avoid race conditions by ensuring that the ``Online`` flag and any
+reboot requests are managed in the same place. The Host controller
+will watch RebootRequest objects and trigger a reconciliation of the linked
+Host when one appears or changes.
+
+For simplicity, we will try to avoid creating a separate controller for the
+RebootRequest at all, but instead perform all operations on it in the course of
+reconciling the BareMetalHost object that it references. However, it may yet
+prove necessary to also create a trivial controller for the purposes of basic
+housekeeping (e.g. setting the owner of the resource).
```

Comment:
That won't be necessary. See my other comment that's "outdated". It doesn't hurt to say we might need this, but I can't think of a reason why we would.

Reply:
That works for me, then.